### ISSUE TITLE
Everything works well except for files

### ISSUE THREAD

**Issue by nigel-dewar** [OfficeDev/O365-Node-Express-Ejs-Sample-App#1]
Hi There, 

This example is really great. Everything works well except for when trying to access files. I get this error"

Whoops! Something has gone wrong.

tunneling socket could not be established, cause=connect ECONNREFUSED

undefined

Error: tunneling socket could not be established, cause=connect ECONNREFUSED
    at ClientRequest.onError (C:\_projects\NextGen\research\O365-Node-Express-Ejs-Sample-App\node_modules\request\node_modules\tunnel-agent\index.js:168:17)
    at ClientRequest.g (events.js:199:16)
    at ClientRequest.emit (events.js:107:17)
    at Socket.socketErrorListener (_http_client.js:272:9)
    at Socket.emit (events.js:107:17)
    at net.js:451:14
    at process._tickCallback (node.js:355:11)

I have set the maximum possible permissions in Azure AD for this app to access SharePoint so should be sweet but not getting any job.

Just another question too if your able to answer it. Is there advantages to using the methods in the example (Passport-azure) vs ADAL.js ?  

Also, it would be really cool if you could update this example hitting a SharePoint list, and demonstrating using ODATA in a query within express api .

Cheers

**Comment by kdeding** [comment]
Hello,
Looks like you enabled proxy by Fiddler, but didn't actually turn on Fiddler. Try start Fiddler while you doing the file. 

Alternatively, you need to set useFiddler = false in the appSetting.js.

Cheers,
Kurt

**Comment by kdeding** [comment]
Hi, 

just realized that I overlooked your question.  I used the passport module because it's node-ready, i.e., it's a proper node module. With ADAL.js, I think you will need to do something yourself to make it a node module. 

I'll follow up on your suggestion on hitting a SharePoint list, when I get a chance.

**Comment by nigel-dewar** [comment]
Hi Mate,

 

Thanks for getting back to me on this. Yeah I tried setting the useFiddler to false, but to no avail, and spent a couple of hours trying different things.

 

In the end I gave up and decided it was high time to dig deep into how all the authentifications systems work because I’ve been trying diffrerent peoples github projects to varying success, but not understanding why some things weren’t working. I was hoping to be lazy, but I guess with this I have will have to understand it inside out.

 

Appreciate your response on this. I was wondering why you used passport, and that makes sense. 

 

Cheers

 

Nigel.

**Comment by nigel-dewar** [comment]
Gday mate,

Just a follow up on this. So after spending a few days fully understanding the authentication pattern using oauth2 etc, I am able to successfully get all the data I want from O365 except for sharepoint list data. Using the new Unified API's I can get access to files.

But SharePoint lists are a completely different kettle of fish. I can get at the data just fine using Fiddler. I think there might be some special headers required to be used when making a call to sharepoint lists. I have read that because SharePoint runs on iis and was not using TLS, but the older SSLv3 (which of course has recently been proven to be insecure), that one may need to specify to node, to use SSLv3 (which it does not by default). Apart from that I'm pretty stumped.

Is there a particular reason you choose request.js for making post,get calls as apposed to the standard node express (module thingy for making post,get) calls?

I'd be really interested to see if you can get sharepoint list data back, because mate I have literally tried everything I can think of, and getting no-where.

Kind Regards

Nigel D.

**Comment by jthake** [comment]
You cannot use the Office 365 Unified APIs to get at SharePoint List Data, like you can Files. You would have to reach out to the /_api/ SharePoint APIs directly to do this. I have written about this here http://www.jeremythake.com/2014/06/using-the-sharepoint-csom-and-rest-api-with-office-365-api-via-azure-ad/

**Comment by nigel-dewar** [comment]
Thanks J, no I realize that. I'm totally accross the fact that there are different end points, and that by obtaining access_tokens from certain endpoints, these can only be used in certain places. The unified endpoints certainly do simplify things, but know for sure that SharePoint is different.

FYI, with Andrew Connell's help, I was able to figure out what the problem actually was. You'd see the discussion we had over this http://www.andrewconnell.com/blog/azure-ad-oauth2-openid-connect#comment-2096739496

And he pointed to this post here: http://blog.mastykarz.nl/econnreset-error-connecting-office-365-nodejs/ 

When you call a SharePoint list from node.js, I needed to specify security protocol in the Get. Like so;

   request.get({
            url:apiUrl,
            agentOptions: {
                secureProtocol: 'TLSv1_method'
            },
            headers:headers,
            json:true
        },function(err,response,messages){ ....

You could put all these options in the get request;

 hostname: "contoso.sharepoint.com",
    path: "/_api/search/query?queryText='size>0'",
    headers: {
        "Authorization": "Bearer " + accessToken,
        "Accept": "application/json;odata=verbose"
    },
    secureOptions: require('constants').SSL_OP_NO_TLSv1_2,
    ciphers: 'ECDHE-RSA-AES256-SHA:AES256-SHA:RC4-SHA:RC4:HIGH:!MD5:!aNULL:!EDH:!AESGCM',
    honorCipherOrder: true

But I found it was not needed, simply the inclusion of;

  agentOptions: {
                secureProtocol: 'TLSv1_method'
            },

did the trick.

Without this, you get no-where. This is not this 'O365-Node-Express-Ejs-Sample-App' and would be helpful for others, because I spent 2 days on this.

So you would be fine hitting all of the other O365 endpoints without including these agentOptions, but when trying to hit SharePoint, you need it.

I suppose most guys would be hitting Sharepoint lists directly from the frontend using ajax, or whatever. But I found very few examples out there accessing Sharepoint lists from the backend in node for some reason.

For me this was important. Because I was following a model that I like from Alex Zanfir, where he hits Google api's, twitter, facebook etc from the backend instead of making direct calls from the frontend. That way he can do some really funky stuff in the backend with those posts,gets, and control them, like sheduling out posts to twitter for instance at a certain time of day. This might not be ideal when making calls from the frontend where its usually getting/posting data instantly, and not really delayed.

Anyway, long story short.

If you update the 'O365-Node-Express-Ejs-Sample-App' with a SharePoint list (GET,POST) example, I'm sure it will be very helpful for some and save them a lot of pain.

Love your work guys. Take care.

**Comment by kdeding** [comment]
Nigel,
Thanks for sharing your findings. The added option does the trick well. :) I'll update the samples to add this for file and list calls. 

After a little digging, I also found out how to get SharePoint list off your O365 sharepoint site. The steps are the same as for the files, except that you'd need to
1. get the SharePoint-specific access token 
2. specify the sharepoint lists resource as "https://<tenant>.sharepoint.com/_api/web/lists", and
3. add the "headers : { 'accept' : 'application/json'}" to the options.

Without 3. you will get a response body in XML, instead of JSON.

Again, I'll include this in my update of the sample. 

Also, I tried to use the node's HTTPS request call, but was not very successful. I couldn't get it to accept the OAuth2 bearer token. But that could be due my ignorance or being lazy. :( Anyway, when I tried it with the request module, it worked. So, that was good enough for me at the time. 

I whole-heartedly agree with you on using node to hit o365 from the backend. It may not be as flashy as the front-end stuff, but it's definitely more "fulfilling". :)

Cheers,
Kurt

**Comment by nigel-dewar** [comment]
Hi Kurt,

 

Thanks for your response, and thanks for taking time out to read my post, and update the github project.

 

I’ve got all the SP list data from my O365  flowing back to me all ready and its good, and overcome all the authentication stuff, like getting the right token for SP endpoint vs the other o365 endpoints. But the updates to this project will save people a lot of pain. In my case the pain was worth it because I learnt so much. I feel like I can do the whole authenticate, and get data process with my eyes closed now.

 

Cheers

 

Nige

**Comment by nigel-dewar** [comment]
The only thing I’d say to be weary of that I have noticed ( and which I am trying to fix now ), is that when doing a get to sharepoint lists, there might need to be some sort of buffering or streaming control or something like that.

 

I say this because I am getting really weird results when doing a get on a sharepoint list with a decent amount of data in it.

 

I’m not sure if this is because my code is bad ( very likely ), or one does need to control the stream.

 

Cheers

**Comment by nigel-dewar** [comment]
Please ignore my last comment. Console.log was pumping out weird data but the res I'm sending back to the front end is just fine. I'm an idiot. Another 2 hours down the drain.


Sent from Samsung Mobile

<div>-------- Original message --------</div><div>From: kdeding <notifications@github.com> </div><div>Date:26/06/2015  06:32  (GMT+08:00) </div><div>To: OfficeDev/O365-Node-Express-Ejs-Sample-App <O365-Node-Express-Ejs-Sample-App@noreply.github.com> </div><div>Cc: nigel-dewar <nigeldewar@live.com> </div><div>Subject: Re: [O365-Node-Express-Ejs-Sample-App] Everything works well except for files (#1) </div><div>
</div>
Nigel,
Thanks for sharing your findings. The added option does the trick well. :) I'll update the samples to add this for file and list calls.

After a little digging, I also found out how to get SharePoint list off your O365 sharepoint site. The steps are the same as for the files, except that you'd need to
1. get the SharePoint-specific access token
2. specify the sharepoint lists resource as "https://<tenant>.sharepoint.com/_api/web/lists", and
3. add the "headers : { 'accept' : 'application/json'}" to the options.

Without 3. you will get a response body in XML, instead of JSON.

Again, I'll include this in my update of the sample.

Also, I tried to use the node's HTTPS request call, but was not very successful. I couldn't get it to accept the OAuth2 bearer token. But that could be due my ignorance or being lazy. :( Anyway, when I tried it with the request module, it worked. So, that was good enough for me at the time.

I whole-heartedly agree with you on using node to hit o365 from the backend. It may not be as flashy as the front-end stuff, but it's definitely more "fulfilling". :)

Cheers,
Kurt

---

**Comment by nigel-dewar** [comment]
You’re the man Kurt!

 

Thanks very much.

**Comment by nigel-dewar** [comment]
Just a note on what you said the other day, about hitting O365 from the node vs the browser, ‘is not as fancy, but is fulfilling’,

I totally agree with you and then some.

 

A lot of people (most) will be hitting O365 direct from the browser, which is fine, and most likely the easiest way to do it. But since we are quickly moving into an IOT world, that’s where hitting these cloud services direct from node becomes more important to me. And as far as I can see, node is the perfect fundamental base for IOT. And it’s being demonstrated in so many contexts now, either using node to control drones ( node copter ) to many other things.

 

SharePoint lists would not seem a likely data store for an IOT device I agree. In IOT you want a high speed data store, where the response time is really low in the milliseconds. So one would rather use a mongo or documentDB to talk to. That being the case though, SharePoint with O365 as a data store offers a level of redundancy and flexibility as a data store that is hard to match. And meets a lot of compliance features that businesses expect. 

 

I’d say, Bill Gates vision in ‘The road ahead’ , of a completely unified world where a person can control, his home, his car, his work, his life very easily through a centralized IT service, is just around the corner. 

 

Thanks Kurt. I really appreciate you updating this project and I hope it helps out a tonne of other people get going with O365.
